File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/model/instructions/EndKaigiInstruction.java
package org.hyperledger.iroha.android.model.instructions;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/** Typed builder for {@code EndKaigi} instructions. */
public final class EndKaigiInstruction implements InstructionTemplate {

  private static final String ACTION = "EndKaigi";

  private final KaigiInstructionUtils.CallId callId;
  private final Long endedAtMs;
  private final Map<String, String> arguments;

  private EndKaigiInstruction(final Builder builder) {
    this(builder, builder.canonicalArguments());
  }

  private EndKaigiInstruction(final Builder builder, final Map<String, String> argumentOrder) {
    this.callId = builder.callId;
    this.endedAtMs = builder.endedAtMs;
    this.arguments = Map.copyOf(argumentOrder);
  }

  public KaigiInstructionUtils.CallId callId() {
    return callId;
  }

  public Long endedAtMs() {
    return endedAtMs;
  }

  @Override
  public InstructionKind kind() {
    return InstructionKind.CUSTOM;
  }

  @Override
  public Map<String, String> toArguments() {
    return arguments;
  }

  public static EndKaigiInstruction fromArguments(final Map<String, String> arguments) {
    final Builder builder = builder();
    builder.setCallId(KaigiInstructionUtils.parseCallId(arguments, "call"));
    final Long ended =
        KaigiInstructionUtils.parseOptionalUnsignedLong(arguments.get("ended_at_ms"), "ended_at_ms");
    builder.setEndedAtMs(ended);
    return new EndKaigiInstruction(builder, new LinkedHashMap<>(arguments));
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (!(obj instanceof EndKaigiInstruction other)) {
      return false;
    }
    return Objects.equals(callId.domainId(), other.callId.domainId())
        && Objects.equals(callId.callName(), other.callId.callName())
        && Objects.equals(endedAtMs, other.endedAtMs);
  }

  @Override
  public int hashCode() {
    return Objects.hash(callId.domainId(), callId.callName(), endedAtMs);
  }

  public static final class Builder {
    private KaigiInstructionUtils.CallId callId;
    private Long endedAtMs;

    private Builder() {}

    public Builder setCallId(final String domainId, final String callName) {
      this.callId = new KaigiInstructionUtils.CallId(domainId, callName);
      return this;
    }

    public Builder setCallId(final KaigiInstructionUtils.CallId callId) {
      this.callId = Objects.requireNonNull(callId, "callId");
      return this;
    }

    public Builder setEndedAtMs(final Long endedAtMs) {
      if (endedAtMs != null && endedAtMs < 0) {
        throw new IllegalArgumentException("endedAtMs must be non-negative");
      }
      this.endedAtMs = endedAtMs;
      return this;
    }

    public EndKaigiInstruction build() {
      if (callId == null) {
        throw new IllegalStateException("callId must be provided");
      }
      return new EndKaigiInstruction(this);
    }

    private Map<String, String> canonicalArguments() {
      final Map<String, String> args = new LinkedHashMap<>();
      args.put("action", ACTION);
      KaigiInstructionUtils.appendCallId(callId, args, "call");
      if (endedAtMs != null) {
        args.put("ended_at_ms", Long.toUnsignedString(endedAtMs));
      }
      return args;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/core/model/instructions/EndKaigiInstruction.kt
package org.hyperledger.iroha.sdk.core.model.instructions

/** Canonical V1 `EndKaigi` wire instruction; private authorization is verified against ledger state. */
class EndKaigiInstruction @JvmOverloads constructor(
    @JvmField val callId: KaigiInstructionUtils.CallId,
    @JvmField val endedAtMs: Long? = null,
    @JvmField val commitment: KaigiAuthorizationScalarV1? = null,
    @JvmField val nullifierDigest: KaigiAuthorizationScalarV1? = null,
    rosterRoot: String? = null,
    @JvmField val proofBase64: String? = null,
) : KaigiWireInstructionV1 {
    @JvmField val rosterRoot: String? = rosterRoot?.let(KaigiInstructionUtils::canonicalRosterRoot)

    init {
        KaigiInstructionUtils.requireAuthorizationArtifacts(commitment, nullifierDigest, this.rosterRoot, proofBase64)
    }

    override val kind: InstructionKind = InstructionKind.CUSTOM
    override val arguments: Map<String, String> by lazy {
        KaigiInstructionUtils.immutableArguments(buildMap {
        put("action", "EndKaigi")
        KaigiInstructionUtils.appendCallId(callId, this, "call")
        endedAtMs?.let { put("ended_at_ms", java.lang.Long.toUnsignedString(it)) }
        commitment?.let { put("commitment.commitment", it.toHex()) }
        nullifierDigest?.let { put("nullifier.digest", it.toHex()) }
        this@EndKaigiInstruction.rosterRoot?.let { put("roster_root", it) }
        proofBase64?.let { put("proof", it) }
        })
    }

    override fun equals(other: Any?): Boolean = other is EndKaigiInstruction && arguments == other.arguments
    override fun hashCode(): Int = arguments.hashCode()

    companion object {
        @JvmStatic
        fun fromArguments(arguments: Map<String, String>): EndKaigiInstruction {
            KaigiInstructionUtils.requireKnownArguments(arguments, setOf(
                "action", "call.domain_id", "call.call_name", "ended_at_ms", "commitment.commitment", "nullifier.digest", "roster_root", "proof",
            ))
            KaigiInstructionUtils.requireAction(arguments, "EndKaigi")
            return EndKaigiInstruction(
                callId = KaigiInstructionUtils.parseCallId(arguments, "call"),
                endedAtMs = KaigiInstructionUtils.parseOptionalUnsignedLong(arguments["ended_at_ms"], "ended_at_ms"),
                commitment = arguments["commitment.commitment"]?.let(KaigiAuthorizationScalarV1::fromArgument),
                nullifierDigest = arguments["nullifier.digest"]?.let(KaigiAuthorizationScalarV1::fromArgument),
                rosterRoot = arguments["roster_root"],
                proofBase64 = arguments["proof"],
            )
        }
    }
}
